File: ios/Earwyrm/Core/Supabase/BlockManager.swift
import Foundation
import Supabase

@Observable
@MainActor
final class BlockManager {
    var blockedUserIds: Set<UUID> = []

    // MARK: - Fetch

    func fetchBlockedUsers(userId: UUID) async {
        do {
            let result: [UserBlock] = try await supabase
                .from("user_blocks")
                .select("id, blocker_id, blocked_user_id, created_at")
                .eq("blocker_id", value: userId.uuidString)
                .execute()
                .value

            blockedUserIds = Set(result.map(\.blockedUserId))
        } catch {
            print("Fetch blocked users error: \(error)")
        }
    }

    // MARK: - Block

    func block(currentUserId: UUID, targetUserId: UUID) async {
        // Optimistic update
        blockedUserIds.insert(targetUserId)
        Haptics.medium()

        do {
            let insert = UserBlockInsert(blockerId: currentUserId, blockedUserId: targetUserId)
            try await supabase
                .from("user_blocks")
                .insert(insert)
                .execute()
        } catch {
            // Handle duplicate key silently
            let desc = String(describing: error)
            if desc.contains("23505") {
                return
            }
            blockedUserIds.remove(targetUserId)
            print("Block user error: \(error)")
        }
    }

    // MARK: - Unblock

    func unblock(currentUserId: UUID, targetUserId: UUID) async {
        // Optimistic update
        blockedUserIds.remove(targetUserId)
        Haptics.light()

        do {
            try await supabase
                .from("user_blocks")
                .delete()
                .eq("blocker_id", value: currentUserId.uuidString)
                .eq("blocked_user_id", value: targetUserId.uuidString)
                .execute()
        } catch {
            blockedUserIds.insert(targetUserId)
            print("Unblock user error: \(error)")
        }
    }

    // MARK: - Query

    func isBlocked(_ userId: UUID) -> Bool {
        blockedUserIds.contains(userId)
    }

    // MARK: - Report Content

    func reportContent(reporterId: UUID, contentType: String, contentId: UUID, reason: String) async -> Bool {
        do {
            let insert = ContentReportInsert(
                reporterId: reporterId,
                contentType: contentType,
                contentId: contentId,
                reason: reason
            )
            try await supabase
                .from("content_reports")
                .insert(insert)
                .execute()

            Haptics.success()
            return true
        } catch {
            let desc = String(describing: error)
            if desc.contains("23505") {
                // Already reported
                return true
            }
            print("Report content error: \(error)")
            return false
        }
    }
}

File: android/app/src/main/java/com/earwyrm/app/core/supabase/BlockManager.kt
package com.earwyrm.app.core.supabase

import com.earwyrm.app.core.model.*
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BlockManager @Inject constructor(private val supabase: SupabaseClient) {
    private val _blockedUserIds = MutableStateFlow<Set<String>>(emptySet())
    val blockedUserIds: StateFlow<Set<String>> = _blockedUserIds.asStateFlow()

    suspend fun fetchBlockedUsers(userId: String) { try { _blockedUserIds.value = supabase.postgrest.from("user_blocks").select { filter { eq("blocker_id", userId) } }.decodeList<UserBlock>().map { it.blockedUserId }.toSet() } catch (_: Exception) { } }
    suspend fun blockUser(blockerId: String, blockedUserId: String) { try { supabase.postgrest.from("user_blocks").insert(UserBlockInsert(blockerId, blockedUserId)); _blockedUserIds.value = _blockedUserIds.value + blockedUserId } catch (_: Exception) { } }
    suspend fun unblockUser(blockerId: String, blockedUserId: String) { try { supabase.postgrest.from("user_blocks").delete { filter { eq("blocker_id", blockerId); eq("blocked_user_id", blockedUserId) } }; _blockedUserIds.value = _blockedUserIds.value - blockedUserId } catch (_: Exception) { } }
    suspend fun reportContent(reporterId: String, contentType: String, contentId: String, reason: String) { try { supabase.postgrest.from("content_reports").insert(ContentReport(reporterId, contentType, contentId, reason)) } catch (_: Exception) { } }
    fun isBlocked(userId: String) = userId in _blockedUserIds.value
}
